[RESTEASY-2366] Introduce RESTEasy integration with WildFly Elytr…

Skyllarr opened the pull request:
…on - AuthenticationClient for Authentication / SSL

issue link: [RESTEASY-2366](https://issues.jboss.org/browse/RESTEASY-2366)

RFE Proposal :  https://github.com/wildfly/wildfly-proposals/pull/189

These are proposed changes for the use of new client SPI that wildfly-elytron will implement.

asoldano commented:
@Skyllarr , I'll review your PR here soon. In the meantime, please rebase it against master: I believe that will solve the SpecialResourceTest failure.

Skyllarr commented:
@asoldano Thank you for the info! I am currently on PTO with limited access to the internet. I will be back in 3 weeks and will rebase then.

Skyllarr commented:
@asoldano I rebased this against master (SpecialResourceTest failure got fixed). Thanks!

Skyllarr commented:
@asoldano @ronsigal I changed the PR branch to 3.10 as requested. All tests are passing for me locally now. Please review and let me know where the ClientConfigProvider SPI  should be located to avoid circular dependency between wildfly-elytron and resteasy project's testsuite. Thanks a lot

Skyllarr commented:
@asoldano I added remaining tests. Please let me know how you would prefer to resolve circular dependency between wildfly-elytron and resteasy's testsuite.

Skyllarr commented:
@ronsigal @asoldano I was not able to reproduce circular dependency. This PR is ready to be reviewed. Thanks!

jimma commented:
@Skyllarr   Looks this will bring circular dependency (resteasy<->elytron) issue. And spit spi module is not an option for Resteasy 3.10 release. We can't create another spi module and change all the dependency graph in Resteasy.  And this is kind of backwards incompatible change, especially for developer with Resteasy dependency.

To resolve this issue, we may need to make  Elytron client only depends jaxrs api. The idea is creating a new ElytronClientConfigration and pass to create the client :       
 ```
Client myclient = ClientBuilder.newBuilder().build();
      Configuration elytronConfiguraion = new ElytronClientConfiguration(myclient.getConfiguration());
      elytronConfiguraion.register(new ClientConfigProviderFilter(serviceLoaderIterator.next()), Priorities.AUTHENTICATION);
      elytronConfiguraion.register(SecurityContextResolver.class);
      Client client = ClientBuilder.newClient(elytronConfiguraion);
```

The SecurityContextResolve can be something like : 

```
 @Provider
public class SecurityContextResolver implements ContextResolver<Function<String, String>> {
    ...
   @Override
    public Function<String, String> getContext(Class<?> type) {
        if (type == SSLContext.class) {
            return getSslContextFunction;
        }
        return null;
    }
}
```

In ClientHttpEngineBuilder , this can be retrieved by this line check :
```
 if (that.getConfiguration().getRuntimeType()== RuntimeType.CLIENT && that.getProviderFactory().getContextResolver(SslContxt.class, MediaType.WILDF))
```

Or we can simply put this function to configuration property, and get it in ClientHttpEngineBuilder to create the ssl connection.

Skyllarr commented:
> @Skyllarr Looks this will bring circular dependency (resteasy<->elytron) issue. And spit spi module is not an option for Resteasy 3.10 release. We can't create another spi module and change all the dependency graph in Resteasy. And this is kind of backwards incompatible change, especially for developer with Resteasy dependency.

@jimma Why is this not an option? This would be just 1 simple repo with single interface in it. There is no issue with backwards compatibility because this interface was not used before. Both resteasy and elytron would be dependent on this new spi repo which would resolve circular dep. Btw. circular dependency is only in test scope, which is not transitive.

> To resolve this issue, we may need to make Elytron client only depends jaxrs api...

The desscribed approach is not automatic and would require user to use Elytron classes. This RFE was meant in a way that RESTEasy would load elytron configuration automatically.

jimma commented:
> > @Skyllarr Looks this will bring circular dependency (resteasy<->elytron) issue. And spit spi module is not an option for Resteasy 3.10 release. We can't create another spi module and change all the dependency graph in Resteasy. And this is kind of backwards incompatible change, especially for developer with Resteasy dependency.
> 
> @jimma Why is this not an option? This would be just 1 simple repo with single interface in it. There is no issue with backwards compatibility because this interface was not used before. Both resteasy and elytron would be dependent on this new spi repo which would resolve circular dep. Btw. circular dependency is only in test scope, which is not transitive.
>
Add another spi maven module or repo is required add another dependency when upgrade resteasy.
User has to add this dependency, jar file to the app when upgrade this minor version.  

> > To resolve this issue, we may need to make Elytron client only depends jaxrs api...
> 
> The desscribed approach is not automatic and would require user to use Elytron classes. This RFE was meant in a way that RESTEasy would load elytron configuration automatically.

Now client request filter and ssl security info has to be configured manually to client. We can look if this can be configured automatically in the 4.x version.

Skyllarr commented:
@jimma Seems that wildfly-elytron could safely be removed from resteasy dependencies and bom, thus there is no circular dependency anymore. I am reevaluating my tests. Let me know if the interface looks good to you

jimma commented:
Thanks @Skyllarr . So you are going to fix this by remove wildfly-elytron dependency for test ?

Skyllarr commented:
@jimma Yes, I have removed elytron dependency from RESTEasy testuite. Please let me know if this approach is alright. I have also added 2 tests that use `mocked ClientConfigProviderImpl` to verify that the new` ClientConfigProvider` interface works for SSLContext and HTTP BASIC auth. Let me know if I should include more tests with this approach. Thanks.

Skyllarr commented:
@jimma Currently the tests create a separate jar that contains mocked service provider `ClientConfigProviderImplMocked` that is registered in jar's `META-INF/services/` and Main class that uses resteasy client to connect to secured server. Tests verify that client used sslcontext and credentials from mocked provider and connected to secured server successfully. The reason I had to create separate jar is so the other testing classes do not pick up the `META-INF/services/ClientConfigProviderImplMocked`.

We think that tests testing integration with Elytron specifically should be moved to wildfly repository. WDYT?

Skyllarr commented:
I have moved integration tests to wildfly repository

jimma commented:
> @jimma Currently the tests create a separate jar that contains mocked service provider `ClientConfigProviderImplMocked` that is registered in jar's `META-INF/services/` and Main class that uses resteasy client to connect to secured server. Tests verify that client used sslcontext and credentials from mocked provider and connected to secured server successfully. The reason I had to create separate jar is so the other testing classes do not pick up the `META-INF/services/ClientConfigProviderImplMocked`. Is this approach OK?
> 
> We think that tests testing integration with Elytron specifically should be moved to wildfly repository. WDYT?

@Skyllarr  This looks good to me.  You must already check the module dependency and maven dependency when both Resteasy and ELytron upgrade in WildFly, there is no issue with this approach, right ?

Skyllarr commented:
> @Skyllarr This looks good to me. You must already check the module dependency and maven dependency when both Resteasy and ELytron upgrade in WildFly, there is no issue with this approach, right ?

@jimma  I don't know if there is no issue with this approach for sure, but I did manage to locally build WildFly with both RESTEasy and Elytron upgrade built on my feature branches, if that answers your question.

Skyllarr commented:
@ronsigal @asoldano I rebased this on 3.13. 

The CI is failing on MicroProfile TCK timeout tests, but these test are failing for me locally even without my changes.

ronsigal commented:
Hi @Skyllarr,

It will take me a little while to get up to speed on this stuff, but I'll start studying. I wouldn't worry about the MP TCK timeouts. I had the same problem, and I was able to eliminate the problem by extending the timeout.

-Ron

ronsigal commented on resteasy-client/src/main/java/org/jboss/resteasy/client/jaxrs/internal/ClientConfigProviderFilter.java:
```diff
@@ -0,0 +1,42 @@
+    public void filter(ClientRequestContext requestContext) throws IOException {
+        if (requestContext.getHeaderString(HttpHeaders.AUTHORIZATION) == null) {
+            URI uri = requestContext.getUri();
+            String token = clientConfigProvider.getBearerToken(uri);
```
ClientConfigProvider doesn't have any javadoc that guarantees that it will behave nicely if uri is null. Could there be a circumstance in which url == null? Maybe this is a stretch, but what if an application has a ClientRequestFilter that supplies the target URL, and that filter runs after this one? I don't know. But maybe ClientConfigProvider just needs some javadoc that defines its behavior. WDYT, @Skyllarr ?

Skyllarr replied:
@ronsigal Thank you for review, this is a very good point. I thought that requestContext always contains uri that is not null when filters are being evaluated, but that was probably wrong assumption. We could throw an exception of continue silently if uri is null here. 

I think it is best to log it as warning and continue silently to avoid problems in case elytron configuration is present on classpath for other clients that use it.

ronsigal commented on resteasy-client/src/main/java/org/jboss/resteasy/client/jaxrs/internal/ClientConfigProviderFilter.java:
```diff
@@ -0,0 +1,42 @@
+                String password = clientConfigProvider.getPassword(uri);
+                if (username != null && password != null) {
+                    requestContext.getHeaders().putSingle(HttpHeaders.AUTHORIZATION, BasicAuthHelper.createHeader(username, password));
+                }
```
Should it be considered an error if a ClientConfigProvider exists but it doesn't hold either a token or username/password? Or maybe there whould be a WARNING?

Skyllarr replied:
@ronsigal Thanks for review! It is not considered error or warning if ClientConfigProvider does not hold token or credentials. It is not necessary to provide those, it could be that it is present only to provide SSL context.

ronsigal commented on resteasy-dependencies-bom/pom.xml:
```diff
@@ -881,11 +881,6 @@
-                <groupId>org.wildfly.security</groupId>
-                <artifactId>wildfly-elytron</artifactId>
-                <version>${version.org.wildfly.security.wildfly-elytron}</version>
-            </dependency>
```
I guess <version.org.wildfly.security.wildfly-elytron> could be removed ... .

Skyllarr replied:
@ronsigal Thank you. Fixed

ronsigal commented:
I haven't read every test, but it looks like a pretty rich set of tests.

Skyllarr commented on resteasy-client/src/main/java/org/jboss/resteasy/client/jaxrs/ClientHttpEngineBuilder43.java:
```diff
@@ -75,6 +83,7 @@ public ClientHttpEngine build()
       {
          SSLConnectionSocketFactory sslsf = null;
          SSLContext theContext = that.sslContext;
+         Iterator clientConfigProviderIterator = ServiceLoader.load(ClientConfigProvider.class).iterator();
```
@asoldano in most scenarios it would work, but it is possible to have keystore configured without truststore as well. For example in elytron client we have a test configurations that use key-store but not truststore, eg. here: https://github.com/wildfly-security/wildfly-elytron/blob/1.x/auth/client/src/test/resources/org/wildfly/security/auth/client/test-wildfly-config-v1_5.xml#L51 .

Skyllarr commented on testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/security/testjar/ClientConfigTestMainClass.java:
```diff
@@ -0,0 +1,61 @@
+        //CHECKSTYLE.OFF: RegexpSinglelineJava
+        System.out.println(result);
+        //CHECKSTYLE.ON: RegexpSinglelineJava
+    }
```
@asoldano Thank you. Fixed

Skyllarr commented:
@jimma This RFE is now pre-checked. Please let me know if there is anything more needed from me or this can be merged. Thanks
